README.md
# clicktypes

Additional types for the click module that I find useful

CHANGELOG.md
# Changelog

Python `clicktypes` changelog.

## [0.0.7] - 2021-08-31

- Fix changelog dates
- Add support for Click 8.0 (resolves issue "AttributeError: 'NoneType' object has no attribute 'upper'")

## [0.0.6] - 2021-03-23

- Fix markdown links

## [0.0.5] - 2021-03-23

- Add IPStringParam, IPv4StringParam and IPv6StringParam
- Add IPNetworkStringParam, IPv4NetworkStringParam and IPv6NetworkStringParam
- Add basic tests
- Remove strict validation for network parameters
- Reformat files with `black`

## [0.0.4] - 2021-03-20

- Add UrlParam
- Add missing dependancy on 'validators'

## [0.0.3] - 2021-03-20

- Add DomainParam
- Add EmailParam

## [0.0.2] - 2021-03-20

- [Add IP address params](clicktypes/network/README.md):
  - IPAddressParam: IPv4 or IPv6 IP addresses
  - IPv4AddressParam: IPv4 IP addresses
  - IPv6AddressParam: IPv6 IP addresses
- Update main README

## [0.0.1] - 2020-03-19

- Initial release
- [Add IP networking params](clicktypes/network/README.md):
  - IPNetworkParam: IPv4 or IPv6 network addresses
  - IPv4NetworkParam: IPv4 network addresses
  - IPv6NetworkParam: IPv6 network addresses


tests/network_network.py
import click
import clicktypes.network
from ipaddress import IPv4Address, IPv6Address, IPv4Network, IPv6Network
from typing import Tuple, List, Union


@click.command()
@click.argument(
    "networks",
    nargs=-1,
    required=True,
    type=clicktypes.network.IPNetworkParam(),
)
def ip_network_param(networks=Tuple):
    ## Put networks into a unique list
    networks: List[Union[IPv4Network, IPv6Network]] = list(dict.fromkeys(networks))

    ## Print each network result
    for network in networks:
        click.echo(f"{network}")


tests/test_network_network.py
import pytest
from click.testing import CliRunner
from .network_network import ip_network_param
from typing import List


def test_ip_network_param_single_ipv4_valid():
    """Test accepting a single valid IPv4 address (which will be normalized into a prefix)"""
    runner = CliRunner()
    result = runner.invoke(ip_network_param, ["192.0.2.0"])
    assert not result.exception
    assert result.exit_code == 0
    assert result.output.strip() == "192.0.2.0/32"


def test_ip_network_param_single_ipv4_invalid():
    """Test rejecting a single invalid IPv4 address"""
    runner = CliRunner()
    result = runner.invoke(ip_network_param, ["192.0.2.256"])
    assert result.exception
    assert result.exit_code == 2


def test_ip_network_param_single_ipv6_valid():
    """Test accepting a single valid IPv6 address (which will be normalized into a prefix)"""
    runner = CliRunner()
    result = runner.invoke(ip_network_param, ["2001:DB8:0000::0000"])
    assert not result.exception
    assert result.exit_code == 0
    assert result.output.strip() == "2001:db8::/128"


def test_ip_network_param_single_ipv6_invalid():
    """Test rejeting a single invalid IPv6 address"""
    runner = CliRunner()
    result = runner.invoke(ip_network_param, ["2001:DB8:gah::/32"])
    assert result.exception
    assert result.exit_code == 2


def test_ip_network_param_multiple_valid():
    """Test accepting multiple valid networks for both IPv4 and IPv6 with a variety of prefix lengths"""
    ## List of networks to test
    networks: List[str] = [
        "192.0.2.0",
        "192.0.2.0/24",
        "192.0.2.2/28",
        "2001:DB8::/32",
        "2001:DB8:0000::/48",
        "2001:DB8:0101::/33",
    ]

    ## Invoke runner
    runner = CliRunner()
    result = runner.invoke(ip_network_param, networks)

    ## Test for exception and exit code
    assert not result.exception
    assert result.exit_code == 0

    ## Split output into list
    output: List[str] = result.output.splitlines()

    ## Make sure the number of results input matches output
    assert len(output) == len(networks)

    ## Test the results match what is expected
    assert "192.0.2.0/32" in output
    assert "192.0.2.0/24" in output
    assert "192.0.2.0/28" in output
    assert "2001:db8::/32" in output
    assert "2001:db8::/48" in output
    assert "2001:db8::/33" in output


def test_ip_network_param_multiple_invalid():
    """Test rejecting a list of networks which contains a single invalid entry"""
    ## List of networks to test
    networks: List[str] = [
        "192.0.2.0",
        "192.0.2.0/24",
        "192.0.2.2/28",
        "2001:DB8:gah:/32",
    ]

    ## Invoke runner
    runner = CliRunner()
    result = runner.invoke(ip_network_param, networks)

    ## Test for exception and exit code
    assert result.exception
    assert result.exit_code == 2


clicktypes/web/email.py
# -*- coding: utf-8 -*-

"""

Click Parameter Types - Domains

"""

from click import ParamType
from validators import email as email_validator

class EmailParam(ParamType):
    """Validate the parameter is an email address. The email address will be converted to lower case and returned if valid.

    Examples:
        'tEsT@exAmple.com' becomes 'test@example.com'
    """

    def convert(self, value: str, param, context) -> str:
        """The function which will perform validation or normalization

        Arguments:
            value (str): The email address

        Returns:
            str: The normalized and validated email address
        """
        ## Lower case email
        email = f'{value.lower()}'

        ## Check if the email is valid
        if not email_validator(email):
            self.fail(f'Could not validate "{value!r}" as an email address')

        ## Return the lower case domain
        return email

clicktypes/network/address.py
# -*- coding: utf-8 -*-

"""

Click Parameter Types - IP Addresses

These parameters are related to IPv4 and IPv6 addresses.

"""

from click import ParamType
from ipaddress import ip_address, IPv4Address, IPv6Address
from typing import Union

class IPAddressParam(ParamType):
    """Validate the parameter is an IPv4/IPv6 address and normalize into an IPv4Address or IPv6Address object.

    Examples:
        '192.0.2.0' will become IPv4Address('192.0.2.0')
        '2001:0DB8:0004:3f:00::2' will become IPv6Address('2001:db8:4:3f::2')
    """

    def convert(self, value: str, param, context) -> Union[IPv4Address, IPv6Address]:
        """The function which will perform validation or normalization

        Arguments:
            value (str): The IPv4/IPv6 address

        Returns:
            Union[IPv4Address, IPv6Address]: The IPv4Address or IPv6Address object
        """
        ## Attempt address creation
        try:
            address: Union[IPv4Address, IPv6Address] = ip_address(value)
        except ValueError:
            ## Not a valid IPv4/IPv6 address
            self.fail(f'Could not validate "{value!r}" as an IPv4 or IPv6 IP address')
        except Exception as e:
            ## Other exception types should be handled
            self.fail(f'Exception validating "{value!r}" as an IPv4 or IPv6 IP address: {e}', param, context)
        else:
            return address

class IPv4AddressParam(ParamType):
    """Validate the parameter is an IPv4 address and normalize into an IPv4Address object.

    Examples:
        '192.0.2.0' will become IPv4Address('192.0.2.0')
    """

    def convert(self, value: str, param, context) -> IPv4Address:
        """The function which will perform validation or normalization

        Arguments:
            value (str): The IPv4 address

        Returns:
            IPv4Address: The IPv4Address object
        """
        ## Attempt address creation
        try:
            address: IPv4Address = IPv4Address(value)
        except ValueError:
            ## Not a valid IPv4 address
            self.fail(f'Could not validate "{value!r}" as an IPv4 IP address')
        except Exception as e:
            ## Other exception types should be handled
            self.fail(f'Exception validating "{value!r}" as an IPv4 IP address: {e}', param, context)
        else:
            return address

class IPv6AddressParam(ParamType):
    """Validate the parameter is an IPv6 address and normalize into an IPv6Address object.

    Examples:
        '2001:0DB8:0004:3f:00::2' will become IPv6Address('2001:db8:4:3f::2')
    """

    def convert(self, value: str, param, context) -> IPv6Address:
        """The function which will perform validation or normalization

        Arguments:
            value (str): The IPv6 address

        Returns:
            IPv6Address: The IPv6Address object
        """
        ## Attempt address creation
        try:
            address: IPv6Address = IPv6Address(value)
        except ValueError:
            ## Not a valid IPv6 address
            self.fail(f'Could not validate "{value!r}" as an IPv6 IP address')
        except Exception as e:
            ## Other exception types should be handled
            self.fail(f'Exception validating "{value!r}" as an IPv6 IP address: {e}', param, context)
        else:
            return address

class IPStringParam(ParamType):
    """Validate the parameter is an IPv4/IPv6 address, normalize and return as a string.

    Examples:
        '192.0.2.0' will become '192.0.2.0'
        '2001:0DB8:0004:3f:00::2' will become '2001:db8:4:3f::2'
    """

    def convert(self, value: str, param, context) -> str:
        """The function which will perform validation or normalization

        Arguments:
            value (str): The IPv4/IPv6 address

        Returns:
            str: The validated and normalized IP address
        """
        ## Attempt address creation
        try:
            address: Union[IPv4Address, IPv6Address] = ip_address(value)
        except ValueError:
            ## Not a valid IPv4/IPv6 address
            self.fail(f'Could not validate "{value!r}" as an IPv4 or IPv6 IP address')
        except Exception as e:
            ## Other exception types should be handled
            self.fail(f'Exception validating "{value!r}" as an IPv4 or IPv6 IP address: {e}', param, context)

        ## Convert now normalized IP address to string
        address: str = f'{address}'

        ## Return normalized IP address
        return address

class IPv4StringParam(ParamType):
    """Validate the parameter is an IPv4 address, normalize and return as a string.

    Examples:
        '192.0.2.0' will become '192.0.2.0'
    """

    def convert(self, value: str, param, context) -> str:
        """The function which will perform validation or normalization

        Arguments:
            value (str): The IPv4 address

        Returns:
            str: The validated and normalized IP address
        """
        ## Attempt address creation
        try:
            address: IPv4Address = IPv4Address(value)
        except ValueError:
            ## Not a valid IPv4 address
            self.fail(f'Could not validate "{value!r}" as an IPv4 IP address')
        except Exception as e:
            ## Other exception types should be handled
            self.fail(f'Exception validating "{value!r}" as an IPv4 IP address: {e}', param, context)

        ## Convert now normalized IP address to string
        address: str = f'{address}'

        ## Return normalized IP address
        return address

class IPv6StringParam(ParamType):
    """Validate the parameter is an IPv6 address, normalize and return as a string.

    Examples:
        '2001:0DB8:0004:3f:00::2' will become '2001:db8:4:3f::2'
    """

    def convert(self, value: str, param, context) -> str:
        """The function which will perform validation or normalization

        Arguments:
            value (str): The IPv6 address

        Returns:
            str: The validated and normalized IP address
        """
        ## Attempt address creation
        try:
            address: IPv6Address = IPv6Address(value)
        except ValueError:
            ## Not a valid IPv6 address
            self.fail(f'Could not validate "{value!r}" as an IPv6 IP address')
        except Exception as e:
            ## Other exception types should be handled
            self.fail(f'Exception validating "{value!r}" as an IPv6 IP address: {e}', param, context)

        ## Convert now normalized IP address to string
        address: str = f'{address}'

        ## Return normalized IP address
        return address

clicktypes/network/network.py
# -*- coding: utf-8 -*-

"""

Click Parameter Types - Network Prefixes

These parameters are related to IPv4 and IPv6 network prefixes.

"""

from click import ParamType
from ipaddress import IPv4Network, IPv6Network
from typing import Union

class IPNetworkParam(ParamType):
    """Validate the parameter is an IPv4/IPv6 address or prefix and normalize into an IPv4Network or IPv6Network object.

    Examples:
        '192.0.2.0' will become IPv4Network('192.0.2.0/32')
        '2001:0DB8:0004:3f:00::2' will become IPv6Network('2001:db8:4:3f::2/128')
        '192.168.1.1/24' will become IPv4Network('192.0.2.0/24')
    """

    def convert(self, value: str, param, context) -> Union[IPv4Network, IPv6Network]:
        """The function which will perform validation or normalization

        Arguments:
            value (str): The IPv4/IPv6 address or network

        Returns:
            Union[IPv4Network, IPv6Network]: The IPv4Network or IPv6Network object
        """
        ## First attempt to create as IPv4 network
        try:
            network: IPv4Network = IPv4Network(value, strict = False)
        except ValueError:
            ## Does not appear to be an IPv4 IP or network; retry with IPv6
            pass
        except Exception as e:
            ## Other exception types should be handled
            self.fail(f'Exception validating "{value!r}" as an IPv4 IP or network: {e}', param, context)
        else:
            return network

        ## Now try IPv6
        try:
            network: IPv6Network = IPv6Network(value, strict = False)
        except ValueError:
            ## Could not be validated as an IPv4 or IPv6 IP or network
            self.fail(f'Could not validate "{value!r}" as an IPv4 or IPv6 IP or network')
        except Exception as e:
            ## Other exception types should be handled
            self.fail(f'Exception validating "{value!r}" as an IPv6 IP or network: {e}', param, context)
        else:
            return network

class IPv4NetworkParam(ParamType):
    """Validate the parameter is an IPv4 address or prefix and normalize into an IPv4Network object.

    Examples:
        '192.0.2.0' will become IPv4Network('192.0.2.0/32')
        '192.168.1.1/24' will become IPv4Network('192.0.2.0/24')
    """

    def convert(self, value: str, param, context) -> IPv4Network:
        """The function which will perform validation or normalization

        Arguments:
            value (str): The IPv4 address or network

        Returns:
            IPv4Network: The IPv4Network object
        """
        ## Attempt IPv4Network object creation
        try:
            network: IPv4Network = IPv4Network(value, strict = False)
        except ValueError:
            self.fail(f'Could not validate "{value!r}" as an IPv4 IP or network')
        except Exception as e:
            ## Other exception types should be handled
            self.fail(f'Exception validating "{value!r}" as an IPv4 IP or network: {e}', param, context)
        else:
            return network

class IPv6NetworkParam(ParamType):
    """Validate the parameter is an IPv6 address or prefix and normalize into an IPv6Network object.

    Examples:
        '2001:0DB8:0004:3f:00::2' will become IPv6Network('2001:db8:4:3f::2/128')
    """

    def convert(self, value: str, param, context) -> IPv6Network:
        """The function which will perform validation or normalization

        Arguments:
            value (str): The IPv6 address or network

        Returns:
            IPv6Network: The IPv6Network object
        """
        ## Attempt IPv4Network object creation
        try:
            network: IPv6Network = IPv6Network(value, strict = False)
        except ValueError:
            self.fail(f'Could not validate "{value!r}" as an IPv6 IP or network')
        except Exception as e:
            ## Other exception types should be handled
            self.fail(f'Exception validating "{value!r}" as an IPv6 IP or network: {e}', param, context)
        else:
            return network

class IPNetworkStringParam(ParamType):
    """Validate the parameter is an IPv4/IPv6 address or prefix, normalize and return value as a string.

    Examples:
        '192.0.2.0' will become '192.0.2.0/32'
        '2001:0DB8:0004:3f:00::2' will become '2001:db8:4:3f::2/128'
        '192.168.1.1/24' will become '192.0.2.0/24'
    """

    def convert(self, value: str, param, context) -> str:
        """The function which will perform validation or normalization

        Arguments:
            value (str): The IPv4/IPv6 address or network

        Returns:
            str: The network prefix represented as a string.
        """
        ## First attempt to create as IPv4 network
        try:
            network: IPv4Network = IPv4Network(value, strict = False)
        except ValueError:
            ## Does not appear to be an IPv4 IP or network; retry with IPv6
            network: None
        except Exception as e:
            ## Other exception types should be handled
            self.fail(f'Exception validating "{value!r}" as an IPv4 IP or network: {e}', param, context)

        ## Now try IPv6
        if not network:
            try:
                network: IPv6Network = IPv6Network(value, strict = False)
            except ValueError:
                ## Could not be validated as an IPv4 or IPv6 IP or network
                self.fail(f'Could not validate "{value!r}" as an IPv4 or IPv6 IP or network')
            except Exception as e:
                ## Other exception types should be handled
                self.fail(f'Exception validating "{value!r}" as an IPv6 IP or network: {e}', param, context)

        ## Convert network object into string
        network: str = f'{network}'

        ## Return the now normalized network
        return network

class IPv4NetworkStringParam(ParamType):
    """Validate the parameter is an IPv4 address or prefix, normalize and return value as a string.

    Examples:
        '192.0.2.0' will become '192.0.2.0/32'
        '192.168.1.1/24' will become '192.0.2.0/24'
    """

    def convert(self, value: str, param, context) -> str:
        """The function which will perform validation or normalization

        Arguments:
            value (str): The IPv4 address or network

        Returns:
            str: The network prefix represented as a string.
        """
        ## Attempt creation of IPv4Network object
        try:
            network: IPv4Network = IPv4Network(value, strict = False)
        except ValueError:
            ## Does not appear to be an IPv4 IP or network
            self.fail(f'Could not validate "{value!r}" as an IPv4 IP or network')
        except Exception as e:
            ## Other exception types should be handled
            self.fail(f'Exception validating "{value!r}" as an IPv4 IP or network: {e}', param, context)

        ## Convert network object into string
        network: str = f'{network}'

        ## Return the now normalized network
        return network

class IPv6NetworkStringParam(ParamType):
    """Validate the parameter is an IPv6 address or prefix, normalize and return value as a string.

    Examples:
        '2001:0DB8:0004:3f:00::2' will become '2001:db8:4:3f::2/128'
    """

    def convert(self, value: str, param, context) -> str:
        """The function which will perform validation or normalization

        Arguments:
            value (str): The IPv6 address or network

        Returns:
            str: The network prefix represented as a string.
        """
        ## Attempt creation of IPv4Network object
        try:
            network: IPv6Network = IPv6Network(value, strict = False)
        except ValueError:
            ## Does not appear to be an IPv4 IP or network
            self.fail(f'Could not validate "{value!r}" as an IPv6 IP or network')
        except Exception as e:
            ## Other exception types should be handled
            self.fail(f'Exception validating "{value!r}" as an IPv6 IP or network: {e}', param, context)

        ## Convert network object into string
        network: str = f'{network}'

        ## Return the now normalized network
        return network